app/api/csv/importEntity.ts

```diff
@@ -66,6 +71,120 @@ type Options = {
   language: string;
 };
 
+const filterJSObject = (input: { [k: string]: any }, keys: string[]): { [k: string]: any } => {
+  const result: { [k: string]: any } = {};
+  keys.forEach(k => {
+    if (input.hasOwnProperty(k)) {
+      result[k] = input[k];
+    }
+  });
+  return result;
+};
+
+const arrangeThesauri = async (
+  file: ImportFile,
+  template: TemplateSchema,
+  languages?: string[],
+  errorContext?: any
+) => {
+  let nameToThesauriIdSelects: { [k: string]: string } = {};
+  let nameToThesauriIdMultiselects: { [k: string]: string } = {};
+  const thesauriIdToExistingValues = new Map();
+  const thesauriIdToNewValues: Map<string, Set<string>> = new Map();
+  const thesauriIdToNormalizedNewValues = new Map();
+  const thesauriRelatedProperties = template.properties?.filter(p =>
+    ['select', 'multiselect'].includes(p.type)
+  );
+  thesauriRelatedProperties?.forEach(p => {
+    if (p.content && p.type) {
+      const thesarusID = p.content.toString();
+      if (p.type === propertyTypes.select) {
+        nameToThesauriIdSelects[p.name] = thesarusID;
+        languages?.forEach(suffix => {
+          nameToThesauriIdSelects[`${p.name}__${suffix}`] = thesarusID;
+        });
+      } else if (p.type === propertyTypes.multiselect) {
+        nameToThesauriIdMultiselects[p.name] = thesarusID;
+        languages?.forEach(suffix => {
+          nameToThesauriIdMultiselects[`${p.name}__${suffix}`] = thesarusID;
+        });
+      }
+    }
+  });
+  const allRelatedThesauri = await Promise.all(
+    Array.from(
+      new Set(thesauriRelatedProperties?.map(p => p.content?.toString()).filter(t => t))
+    ).map(async id => thesauri.getById(id))
+  );
+  allRelatedThesauri.forEach(t => {
+    if (t) {
+      const id = t._id.toString();
+      thesauriIdToExistingValues.set(
+        id,
+        new Set(t.values?.map(v => normalizeThesaurusLabel(v.label)))
+      );
+      thesauriIdToNewValues.set(id, new Set());
+      thesauriIdToNormalizedNewValues.set(id, new Set());
+    }
+  });
+  function handleLabels(id: string, original: string, normalized: string | null) {
+    if (
+      normalized &&
+      !thesauriIdToExistingValues.get(id).has(normalized) &&
+      !thesauriIdToNormalizedNewValues.get(id).has(normalized)
+    ) {
+      thesauriIdToNewValues.get(id)?.add(original);
+      thesauriIdToNormalizedNewValues.get(id).add(normalized);
+    }
+  }
+  await csv(await file.readStream(), errorContext?.stopOnError)
```

Comment:
@LaszloKecskes errorContext is confusing, its being used as error "emiter" and in this case its used to grab an option, and it actually its not clear what this object represents, it is the parent, csvLoader, i think its better if we do not pass the whole csvLoader, but instead the things we need here as parameters, like options for the stopOnError.

for the errors, can we make this handle its own errors and just throw them, the csvLoader can handle them in whatever manner.

this change basically means that this function should know nothing about the parent, throw errors in case it has and receive params like stopOnError, what do you think ?

Reply:
Done. I'm not confident about how the load function handles the errors thrown by the arrangeThesauri, we could probably doublecheck that.

Reply:
ok, just reviewed it, i think its fine but needs to be tested and moved probably.